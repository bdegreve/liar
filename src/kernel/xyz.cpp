/** @file
 *  @author Bram de Greve (bramz@users.sourceforge.net)
 *
 *  LiAR isn't a raytracer
 *  Copyright (C) 2004-2010  Bram de Greve (bramz@users.sourceforge.net)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *  http://liar.bramz.net/
 */

#include "kernel_common.h"
#include "xyz.h"
#include <lass/stde/extended_iterator.h>
#include <lass/stde/access_iterator.h>

namespace liar
{
namespace kernel
{
namespace impl
{

/** load the XYZ sensitivity curves of the CIE-1964 standard observer.
 * 
 *  http://www.isc.tamu.edu/~astro/color/cie_xyz1964.html
 *  The curves are rescaled to have area 1 for all channels.
 */
Observer loadStandardObserver()
{
	const TScalar w[] =
	{
		3.70e-007f,
		3.75e-007f,
		3.80e-007f,
		3.85e-007f,
		3.90e-007f,
		3.95e-007f,
		4.00e-007f,
		4.05e-007f,
		4.10e-007f,
		4.15e-007f,
		4.20e-007f,
		4.25e-007f,
		4.30e-007f,
		4.35e-007f,
		4.40e-007f,
		4.45e-007f,
		4.50e-007f,
		4.55e-007f,
		4.60e-007f,
		4.65e-007f,
		4.70e-007f,
		4.75e-007f,
		4.80e-007f,
		4.85e-007f,
		4.90e-007f,
		4.95e-007f,
		5.00e-007f,
		5.05e-007f,
		5.10e-007f,
		5.15e-007f,
		5.20e-007f,
		5.25e-007f,
		5.30e-007f,
		5.35e-007f,
		5.40e-007f,
		5.45e-007f,
		5.50e-007f,
		5.55e-007f,
		5.60e-007f,
		5.65e-007f,
		5.70e-007f,
		5.75e-007f,
		5.80e-007f,
		5.85e-007f,
		5.90e-007f,
		5.95e-007f,
		6.00e-007f,
		6.05e-007f,
		6.10e-007f,
		6.15e-007f,
		6.20e-007f,
		6.25e-007f,
		6.30e-007f,
		6.35e-007f,
		6.40e-007f,
		6.45e-007f,
		6.50e-007f,
		6.55e-007f,
		6.60e-007f,
		6.65e-007f,
		6.70e-007f,
		6.75e-007f,
		6.80e-007f,
		6.85e-007f,
		6.90e-007f,
		6.95e-007f,
		7.00e-007f,
		7.05e-007f,
		7.10e-007f,
		7.15e-007f,
		7.20e-007f,
		7.25e-007f,
		7.30e-007f,
		7.35e-007f,
		7.40e-007f,
		7.45e-007f,
		7.50e-007f,
		7.55e-007f,
		7.60e-007f,
		7.65e-007f,
		7.70e-007f,
		7.75e-007f,
		7.80e-007f,
	};
	const XYZ xyz[] =
	{
		XYZ(0.0000e000f, 0.0000e000f, 0.0000e000f),
		XYZ(0.0000e000f, 0.0000e000f, 0.0000e000f),
		XYZ(1.7146e003f, 0.0000e000f, 5.9997e003f),
		XYZ(6.0010e003f, 8.5718e002f, 2.4856e004f),
		XYZ(2.0575e004f, 2.5715e003f, 8.9996e004f),
		XYZ(6.1725e004f, 6.8574e003f, 2.7684e005f),
		XYZ(1.6374e005f, 1.7144e004f, 7.3711e005f),
		XYZ(3.7206e005f, 3.8573e004f, 1.6894e006f),
		XYZ(7.2612e005f, 7.5432e004f, 3.3376e006f),
		XYZ(1.2053e006f, 1.2429e005f, 5.6295e006f),
		XYZ(1.7532e006f, 1.8344e005f, 8.3353e006f),
		XYZ(2.2692e006f, 2.5287e005f, 1.0992e007f),
		XYZ(2.6979e006f, 3.3173e005f, 1.3315e007f),
		XYZ(3.0665e006f, 4.2516e005f, 1.5415e007f),
		XYZ(3.2894e006f, 5.3231e005f, 1.6862e007f),
		XYZ(3.3151e006f, 6.4031e005f, 1.7376e007f),
		XYZ(3.1780e006f, 7.6717e005f, 1.7098e007f),
		XYZ(2.9405e006f, 9.1118e005f, 1.6291e007f),
		XYZ(2.5916e006f, 1.0989e006f, 1.4960e007f),
		XYZ(2.1784e006f, 1.3098e006f, 1.3327e007f),
		XYZ(1.6769e006f, 1.5875e006f, 1.1293e007f),
		XYZ(1.1342e006f, 1.8849e006f, 8.8299e006f),
		XYZ(6.9012e005f, 2.1738e006f, 6.6177e006f),
		XYZ(3.5235e005f, 2.5518e006f, 4.8863e006f),
		XYZ(1.3888e005f, 2.9067e006f, 3.5596e006f),
		XYZ(4.3722e004f, 3.3893e006f, 2.5919e006f),
		XYZ(3.2577e004f, 3.9499e006f, 1.8728e006f),
		XYZ(1.3202e005f, 4.5550e006f, 1.3645e006f),
		XYZ(3.2148e005f, 5.2005e006f, 9.5996e005f),
		XYZ(6.1210e005f, 5.8777e006f, 7.0454e005f),
		XYZ(1.0090e006f, 6.5300e006f, 5.2026e005f),
		XYZ(1.4831e006f, 7.0571e006f, 3.6941e005f),
		XYZ(2.0275e006f, 7.5020e006f, 2.6142e005f),
		XYZ(2.6079e006f, 7.9186e006f, 1.7656e005f),
		XYZ(3.2303e006f, 8.2460e006f, 1.1742e005f),
		XYZ(3.8715e006f, 8.4192e006f, 6.7711e004f),
		XYZ(4.5419e006f, 8.5015e006f, 3.4284e004f),
		XYZ(5.2817e006f, 8.5641e006f, 9.4281e003f),
		XYZ(6.0456e006f, 8.5486e006f, 0.0000e000f),
		XYZ(6.8051e006f, 8.4209e006f, 0.0000e000f),
		XYZ(7.5330e006f, 8.1912e006f, 0.0000e000f),
		XYZ(8.1545e006f, 7.8449e006f, 0.0000e000f),
		XYZ(8.6946e006f, 7.4480e006f, 0.0000e000f),
		XYZ(9.2098e006f, 7.0769e006f, 0.0000e000f),
		XYZ(9.5888e006f, 6.6637e006f, 0.0000e000f),
		XYZ(9.7242e006f, 6.1751e006f, 0.0000e000f),
		XYZ(9.6359e006f, 5.6428e006f, 0.0000e000f),
		XYZ(9.3367e006f, 5.0908e006f, 0.0000e000f),
		XYZ(8.8343e006f, 4.5259e006f, 0.0000e000f),
		XYZ(8.1502e006f, 3.9584e006f, 0.0000e000f),
		XYZ(7.3410e006f, 3.4124e006f, 0.0000e000f),
		XYZ(6.4717e006f, 2.9110e006f, 0.0000e000f),
		XYZ(5.5509e006f, 2.4301e006f, 0.0000e000f),
		XYZ(4.5873e006f, 1.9569e006f, 0.0000e000f),
		XYZ(3.7001e006f, 1.5412e006f, 0.0000e000f),
		XYZ(2.9465e006f, 1.2018e006f, 0.0000e000f),
		XYZ(2.3001e006f, 9.2232e005f, 0.0000e000f),
		XYZ(1.7514e006f, 6.9603e005f, 0.0000e000f),
		XYZ(1.3082e006f, 5.1688e005f, 0.0000e000f),
		XYZ(9.6188e005f, 3.7802e005f, 0.0000e000f),
		XYZ(6.9697e005f, 2.7258e005f, 0.0000e000f),
		XYZ(4.9637e005f, 1.9372e005f, 0.0000e000f),
		XYZ(3.5063e005f, 1.3629e005f, 0.0000e000f),
		XYZ(2.4518e005f, 9.5147e004f, 0.0000e000f),
		XYZ(1.7060e005f, 6.6003e004f, 0.0000e000f),
		XYZ(1.1831e005f, 4.6288e004f, 0.0000e000f),
		XYZ(8.2300e004f, 3.1716e004f, 0.0000e000f),
		XYZ(5.6581e004f, 2.2287e004f, 0.0000e000f),
		XYZ(3.9435e004f, 1.5429e004f, 0.0000e000f),
		XYZ(2.6576e004f, 1.0286e004f, 0.0000e000f),
		XYZ(1.8860e004f, 6.8574e003f, 0.0000e000f),
		XYZ(1.2859e004f, 5.1431e003f, 0.0000e000f),
		XYZ(8.5729e003f, 3.4287e003f, 0.0000e000f),
		XYZ(6.0010e003f, 2.5715e003f, 0.0000e000f),
		XYZ(4.2864e003f, 1.7144e003f, 0.0000e000f),
		XYZ(3.4291e003f, 8.5718e002f, 0.0000e000f),
		XYZ(2.5719e003f, 8.5718e002f, 0.0000e000f),
		XYZ(1.7146e003f, 8.5718e002f, 0.0000e000f),
		XYZ(8.5729e002f, 0.0000e000f, 0.0000e000f),
		XYZ(8.5729e002f, 0.0000e000f, 0.0000e000f),
		XYZ(8.5729e002f, 0.0000e000f, 0.0000e000f),
		XYZ(0.0000e000f, 0.0000e000f, 0.0000e000f),
		XYZ(0.0000e000f, 0.0000e000f, 0.0000e000f),
	};

	return Observer(w, w + sizeof w / sizeof(TScalar), xyz);
}

template <typename It>
struct cdf_accessor_t: std::unary_function<It, TScalar>
{
	typedef const TScalar value_type;
	typedef const TScalar reference;
	typedef const TScalar* pointer;
	cdf_accessor_t(const XYZ& weight = XYZ()): weight_(weight) {}
	TScalar operator()(It i) const
	{
		return dot(weight_, i->cdf); 
	}
private:
	XYZ weight_;
};

template <typename It>
stde::access_iterator_t< It, cdf_accessor_t<It> > cdf_iterator(It i, const XYZ& weight)
{
	return stde::access_iterator_t< It, cdf_accessor_t<It> >(i, cdf_accessor_t<It>(weight));
}

}

namespace 
{
	const TFrequency c0 = 299792458; // speed of light in vacuum
}

void Observer::init()
{
	nodes_.front().cdf = XYZ();
	for (TNodes::iterator i = stde::next(nodes_.begin()); i != nodes_.end(); ++i)
	{
		Node& node = *i;
		const Node& prev = *stde::prev(i);
		node.cdf = prev.cdf + prev.xyz * (node.wavelength - prev.wavelength);
	}
	for (TNodes::iterator i = nodes_.begin(); i != stde::prev(nodes_.end()); ++i)
	{
		Node& node = *i;
		const Node& next = *stde::next(i);
		node.dxyz_dw = (next.xyz - node.xyz) / (next.wavelength - node.wavelength);
	}
	nodes_.back().dxyz_dw = XYZ();
}

const Observer::TFrequencyRange Observer::frequencyRange() const
{
	return std::make_pair(c0 / nodes_.front().wavelength, c0 / nodes_.back().wavelength);
}

const XYZ Observer::tristimulus(TFrequency frequency) const
{
	const TScalar wavelength = c0 / frequency;
	typedef const Node TConstNode;
	const TNodes::const_iterator i = std::lower_bound(
		stde::const_member_iterator(nodes_.begin(), &Node::wavelength),
		stde::const_member_iterator(nodes_.end(), &Node::wavelength),
		wavelength).base();
	const Node& node = *(i == nodes_.begin() ? i : stde::prev(i));
	return node.xyz + node.dxyz_dw * (wavelength - node.wavelength);
}

const XYZ Observer::chromaticity(TFrequency frequency) const
{
	return kernel::chromaticity(tristimulus(frequency));
}

TFrequency Observer::sample(const XYZ& weight, TScalar sample, XYZ& chromaticity, TScalar& pdf) const
{
	const XYZ w = kernel::chromaticity(weight);
	TNodes::const_iterator i = std::lower_bound(
		impl::cdf_iterator(nodes_.begin(), w),
		impl::cdf_iterator(nodes_.end(), w),
		sample).base();
	const Node& node = *(i == nodes_.begin() ? i : stde::prev(i));
	const TScalar p = dot(node.xyz, w);
	LASS_ASSERT(p >= 0);
	const TScalar c = dot(node.cdf, w);
	LASS_ASSERT(c <= sample);
	const TScalar dWavelength = p > 0 ? (sample - c) / p : 0;
	const TScalar wavelength = node.wavelength + dWavelength;
	const XYZ xyz = node.xyz + node.dxyz_dw * dWavelength;
	chromaticity = kernel::chromaticity(xyz);	
	pdf = p;
	return c0 / wavelength;
}

const XYZ chromaticity(const XYZ& xyz)
{
	const TScalar sum = xyz.x + xyz.y + xyz.z;
	if (sum == 0)
	{
		return 0;
	}
	return xyz / sum;	
}

const Observer& standardObserver()
{
	static const Observer observer = impl::loadStandardObserver();
	return observer;
}

const Observer& forceLoadingAtStartUp = standardObserver();

const XYZ chromaticity(TFrequency frequency)
{
	return chromaticity(tristimulus(frequency));
}

const XYZ tristimulus(TFrequency frequency)
{
	return standardObserver().tristimulus(frequency);
}

TFrequency sampleFrequency(const XYZ& power, TScalar sample, XYZ& chromaticity, TScalar& pdf)
{
	return standardObserver().sample(power, sample, chromaticity, pdf);
}

}

}

// EOF
