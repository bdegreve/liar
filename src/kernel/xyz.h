/** @file
 *  @author Bram de Greve (bramz@users.sourceforge.net)
 *
 *  LiAR isn't a raytracer
 *  Copyright (C) 2004-2009  Bram de Greve (bramz@users.sourceforge.net)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *  http://liar.bramz.net/
 */

#ifndef LIAR_GUARDIAN_OF_INCLUSION_KERNEL_KERNEL_XYZ_H
#define LIAR_GUARDIAN_OF_INCLUSION_KERNEL_KERNEL_XYZ_H

#include "kernel_common.h"

namespace liar
{
namespace kernel
{

class XYZ
{
public:
	TScalar x;
	TScalar y;
	TScalar z;

	XYZ(): x(0), y(0), z(0) {}
	XYZ(TScalar f): x(f), y(f), z(f) {}
	XYZ(TScalar x, TScalar y, TScalar z): x(x), y(y), z(z) {}

	XYZ& operator+=(const XYZ& other) 
	{
		x += other.x;
		y += other.y;
		z += other.z;
		return *this;
	}
	XYZ& operator-=(const XYZ& other) 
	{
		x -= other.x;
		y -= other.y;
		z -= other.z;
		return *this;
	}
	XYZ& operator*=(const XYZ& other) 
	{
		x *= other.x;
		y *= other.y;
		z *= other.z;
		return *this;
	}
	XYZ& operator/=(const XYZ& other) 
	{
		x /= other.x;
		y /= other.y;
		z /= other.z;
		return *this;
	}
	TScalar total() const 
	{
		return x + y + z;
	}
	TScalar absTotal() const
	{
		return num::abs(x) + num::abs(y) + num::abs(z);
	}
	TScalar average() const
	{
		return total() / 3;
	}
	bool isZero() const
	{
		return x == 0 && y == 0 && z == 0;
	}
	bool operator!() const
	{
		return isZero();
	}
	operator num::SafeBool() const
	{
		return isZero() ? num::safeFalse : num::safeTrue;
	}
};

inline TScalar average(const XYZ& a)
{
	return a.x + a.y + a.z / 3;
}

inline TScalar positiveAverage(const XYZ& a)
{
	return std::max<TScalar>(average(a), 0);
}

inline TScalar dot(const XYZ& a, const XYZ& b)
{
	return a.x * b.x + a.y * b.y + a.z * b.z;
}

inline XYZ operator+(const XYZ& a, const XYZ& b) 
{
	XYZ r(a);
	r += b;
	return r;
}

inline XYZ operator-(const XYZ& a, const XYZ& b) 
{
	XYZ r(a);
	r -= b;
	return r;
}
inline XYZ operator*(const XYZ& a, const XYZ& b) 
{
	XYZ r(a);
	r *= b;
	return r;
}
inline XYZ operator/(const XYZ& a, const XYZ& b) 
{
	XYZ r(a);
	r /= b;
	return r;
}
inline XYZ sqr(const XYZ& a)
{
	return XYZ(a.x * a.x, a.y * a.y, a.z * a.z);
}
inline XYZ pow(const XYZ& a, TScalar b)
{
	return XYZ(num::pow(a.x, b), num::pow(a.y, b), num::pow(a.z, b));
}
inline XYZ clamp(const XYZ& a, TScalar min, TScalar max)
{
	return XYZ(num::clamp(a.x, min, max), num::clamp(a.y, min, max), num::clamp(a.z, min, max));
}
inline XYZ lerp(const XYZ& a, const XYZ& b, TScalar f)
{
	return XYZ(num::lerp(a.x, b.x, f), num::lerp(a.y, b.y, f), num::lerp(a.z, b.z, f));
}
inline bool operator==(const XYZ& a, const XYZ& b)
{
	return a.x == b.x && a.y == b.y && a.z == b.z;
}

class LIAR_KERNEL_DLL Observer
{
public:
	typedef std::pair<TFrequency, TFrequency> TFrequencyRange;

	template <typename InputIteratorWavelength, typename InputIteratorXYZ>
	Observer(InputIteratorWavelength firstW, InputIteratorWavelength lastW, InputIteratorXYZ firstXYZ)
	{
		while (firstW != lastW)
		{
			nodes_.push_back(Node(*firstW++, *firstXYZ++));
		}
		init();
	}

	const TFrequencyRange frequencyRange() const;
	const XYZ tristimulus(TFrequency frequency) const;
	const XYZ chromaticity(TFrequency frequency) const;
	TFrequency sample(const XYZ& power, TScalar sample, XYZ& chromaticity, TScalar& pdf) const;

private:
	struct Node
	{
		TScalar wavelength;
		XYZ xyz;
		XYZ dxyz_dw;
		XYZ cdf;
		Node(TScalar w, XYZ xyz): wavelength(w), xyz(xyz) {}
	};
	typedef std::vector<Node> TNodes;

	void init();

	TNodes nodes_;
};

LIAR_KERNEL_DLL const XYZ chromaticity(const XYZ& xyz);

LIAR_KERNEL_DLL const Observer& standardObserver();
LIAR_KERNEL_DLL const XYZ tristimulus(TFrequency frequency);
LIAR_KERNEL_DLL const XYZ chromaticity(TFrequency frequency);
LIAR_KERNEL_DLL TFrequency sampleFrequency(const XYZ& power, TScalar sample, XYZ& chromaticity, TScalar& pdf);

}
}

namespace lass
{
namespace python
{
template <>
struct PyExportTraits<liar::kernel::XYZ>
{
	static PyObject* build(const liar::kernel::XYZ& v)
	{
		return fromSharedPtrToNakedCast(python::makeTuple(v.x, v.y, v.z));
	}
	static int get(PyObject* obj, liar::kernel::XYZ& v)
	{
		return python::decodeTuple(obj, v.x, v.y, v.z);
	}
};
}
}

#endif

// EOF
