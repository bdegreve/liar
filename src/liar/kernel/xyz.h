/** @file
 *  @author Bram de Greve (bramz@users.sourceforge.net)
 *
 *  LiAR isn't a raytracer
 *  Copyright (C) 2004-2025  Bram de Greve (bramz@users.sourceforge.net)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *  http://liar.bramz.net/
 */

#ifndef LIAR_GUARDIAN_OF_INCLUSION_KERNEL_KERNEL_XYZ_H
#define LIAR_GUARDIAN_OF_INCLUSION_KERNEL_KERNEL_XYZ_H

#include "kernel_common.h"

namespace liar
{
namespace kernel
{

class LASS_SIMD_ALIGN XYZ
{
public:
	typedef LIAR_VALUE TValue;
	typedef LIAR_VALUE TParam;
	typedef num::NumTraits<TValue> TNumTraits;

	TValue x;
	TValue y;
	TValue z;

	XYZ(): x(0), y(0), z(0) {}
	XYZ(TParam f) : x(f), y(f), z(f) {}
	XYZ(TParam x, TParam y, TParam z) : x(x), y(y), z(z) {}

	XYZ& operator+=(const XYZ& other)
	{
		x += other.x;
		y += other.y;
		z += other.z;
		return *this;
	}
	XYZ& operator-=(const XYZ& other)
	{
		x -= other.x;
		y -= other.y;
		z -= other.z;
		return *this;
	}
	XYZ& operator*=(const XYZ& other)
	{
		x *= other.x;
		y *= other.y;
		z *= other.z;
		return *this;
	}
	XYZ& operator/=(const XYZ& other)
	{
		x /= other.x;
		y /= other.y;
		z /= other.z;
		return *this;
	}
	TValue total() const
	{
		return x + y + z;
	}
	TValue absTotal() const
	{
		return num::abs(x) + num::abs(y) + num::abs(z);
	}
	TValue average() const
	{
		return total() / 3;
	}
	TValue maximum() const
	{
		return std::max(x, std::max(y, z));
	}
	bool isZero() const
	{
		return x == 0 && y == 0 && z == 0;
	}
	bool isNaN() const
	{
		return num::isNaN(total());
	}
	bool operator!() const
	{
		return isZero();
	}
	explicit operator bool() const
	{
		return !isZero();
	}
};

inline XYZ abs(const XYZ& a)
{
	return XYZ(num::abs(a.x), num::abs(a.y), num::abs(a.z));
}

inline XYZ max(const XYZ& a, const XYZ& b)
{
	return XYZ(std::max(a.x, b.x), std::max(a.y, b.y), std::max(a.z, b.z));
}

inline XYZ::TValue average(const XYZ& a)
{
	return (a.x + a.y + a.z) / 3;
}

inline XYZ::TValue dot(const XYZ& a, const XYZ& b)
{
	return a.x * b.x + a.y * b.y + a.z * b.z;
}

inline XYZ operator+(const XYZ& a, const XYZ& b)
{
	XYZ r(a);
	r += b;
	return r;
}

inline XYZ operator-(const XYZ& a, const XYZ& b)
{
	XYZ r(a);
	r -= b;
	return r;
}
inline XYZ operator*(const XYZ& a, const XYZ& b)
{
	XYZ r(a);
	r *= b;
	return r;
}
inline XYZ operator/(const XYZ& a, const XYZ& b)
{
	XYZ r(a);
	r /= b;
	return r;
}
inline XYZ sqr(const XYZ& a)
{
	return XYZ(a.x * a.x, a.y * a.y, a.z * a.z);
}
inline XYZ pow(const XYZ& a, XYZ::TParam b)
{
	return XYZ(num::pow(a.x, b), num::pow(a.y, b), num::pow(a.z, b));
}
inline XYZ exp(const XYZ& a)
{
	return XYZ(num::exp(a.x), num::exp(a.y), num::exp(a.z));
}
inline XYZ clamp(const XYZ& a, XYZ::TParam min, XYZ::TParam max)
{
	return XYZ(num::clamp(a.x, min, max), num::clamp(a.y, min, max), num::clamp(a.z, min, max));
}
inline XYZ lerp(const XYZ& a, const XYZ& b, XYZ::TParam f)
{
	return XYZ(num::lerp(a.x, b.x, f), num::lerp(a.y, b.y, f), num::lerp(a.z, b.z, f));
}
inline bool operator==(const XYZ& a, const XYZ& b)
{
	return a.x == b.x && a.y == b.y && a.z == b.z;
}

inline const XYZ chromaticity(const XYZ& xyz)
{
	const XYZ::TValue sum = xyz.x + xyz.y + xyz.z;
	return sum == 0 ? 0 : xyz / sum;
}

LIAR_KERNEL_DLL std::ostream& LASS_CALL operator<<(std::ostream& os, const XYZ& xyz);

}
}

namespace lass
{
namespace python
{
template <>
struct PyExportTraits<liar::kernel::XYZ>
{
	constexpr static const char* py_typing = "tuple[float, float, float]";

	static PyObject* build(const liar::kernel::XYZ& v)
	{
		return fromSharedPtrToNakedCast(python::makeTuple(v.x, v.y, v.z));
	}
	static int get(PyObject* obj, liar::kernel::XYZ& v)
	{
		return python::decodeTuple(obj, v.x, v.y, v.z);
	}
};
}
}

#endif

// EOF
