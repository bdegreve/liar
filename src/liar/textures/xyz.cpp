/** @file
 *  @author Bram de Greve (bramz@users.sourceforge.net)
 *
 *  LiAR isn't a raytracer
 *  Copyright (C) 2004-2025  Bram de Greve (bramz@users.sourceforge.net)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 *  http://liar.bramz.net/
 */

#include "textures_common.h"
#include "xyz.h"

namespace liar
{
namespace textures
{

PY_DECLARE_CLASS_DOC(Xyz, "mixes three textures by the x, y and z context channels")
PY_CLASS_CONSTRUCTOR_3(Xyz, const TTextureRef&, const TTextureRef&, const TTextureRef&);
PY_CLASS_MEMBER_RW_DOC(Xyz, textureA, setTextureA, "first texture")
PY_CLASS_MEMBER_RW_DOC(Xyz, textureB, setTextureB, "second texture")
PY_CLASS_MEMBER_RW_DOC(Xyz, textureC, setTextureC, "third texture")

// --- public --------------------------------------------------------------------------------------

Xyz::Xyz(const TTextureRef& a, const TTextureRef& b, const TTextureRef& c):
	a_(a),
	b_(b),
	c_(c)
{
}



void Xyz::setTextureA(const TTextureRef& a)
{
	a_ = a;
}



void Xyz::setTextureB(const TTextureRef& b)
{
	b_ = b;
}



void Xyz::setTextureC(const TTextureRef& c)
{
	c_ = c;
}



const TTextureRef& Xyz::textureA() const
{
	return a_;
}



const TTextureRef& Xyz::textureB() const
{
	return b_;
}



const TTextureRef& Xyz::textureC() const
{
	return c_;
}



// --- protected -----------------------------------------------------------------------------------

const TPyObjectPtr Xyz::doGetState() const
{
	return python::makeTuple(a_, b_, c_);
}



void Xyz::doSetState(const TPyObjectPtr& state)
{
	python::decodeTuple(state, a_, b_, c_);
}



bool Xyz::doIsChromatic() const
{
	return a_->isChromatic() || b_->isChromatic() || c_->isChromatic();
}



// --- private -------------------------------------------------------------------------------------

const Spectral Xyz::doLookUp(const Sample& sample, const IntersectionContext& context, SpectralType type) const
{
	const TPoint3D& p = context.point();
	const TValue x = num::fractional(static_cast<TValue>(p.x));
	const TValue y = num::fractional(static_cast<TValue>(p.y));
	const TValue z = num::fractional(static_cast<TValue>(p.z));
	return Spectral(x * a_->lookUp(sample, context, SpectralType::Illuminant) + y * b_->lookUp(sample, context, SpectralType::Illuminant) + z * c_->lookUp(sample, context, SpectralType::Illuminant), type);
}



Texture::TValue Xyz::doScalarLookUp(const Sample& sample, const IntersectionContext& context) const
{
	const TPoint3D& p = context.point();
	const TValue x = num::fractional(static_cast<TValue>(p.x));
	const TValue y = num::fractional(static_cast<TValue>(p.y));
	const TValue z = num::fractional(static_cast<TValue>(p.z));
	return x * a_->scalarLookUp(sample, context) + y * b_->scalarLookUp(sample, context) + z * c_->scalarLookUp(sample, context);
}



// --- free ----------------------------------------------------------------------------------------

}

}

// EOF
