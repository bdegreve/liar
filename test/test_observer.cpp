#include "kernel/xyz.h"
#include <lass/num/random.h>
#include <lass/num/distribution.h>
#include <lass/util/environment.h>

int test_observer(int, char*[])
{
	const size_t numSamples = 1000000;

	using namespace liar;
	using namespace lass;

	const kernel::Observer& observer = kernel::standardObserver();

	num::RandomMT19937 rng;
	rng.seed(1234);
	num::DistributionUniform<TScalar, num::RandomMT19937> uniform(rng);

	typedef std::vector<kernel::XYZ> TXYZs;
	TXYZs powers;
	powers.push_back(kernel::XYZ(1, 1, 1));
	powers.push_back(kernel::XYZ(1, 0, 0));
	powers.push_back(kernel::XYZ(0, 1, 0));
	powers.push_back(kernel::XYZ(0, 0, 1));
	powers.push_back(kernel::XYZ(.2f, .5f, .3f));

	typedef std::vector<TFrequency> TBins;
	TBins bins;
	const TFrequency fmin = 3.8e14;
	const TFrequency fmax = 8.2e14;
	const size_t nBins = 10;
	for (size_t k = 0; k < nBins; ++k)
	{
		bins.push_back(fmin + k * (fmax - fmin) / nBins);
	}
	bins.push_back(fmax);

	const std::string path = util::getEnvironment<std::string>("TEST_SOURCE_DIR") + "/test_observer.out";
	std::ifstream reference(path.c_str());

	for (TXYZs::const_iterator power = powers.begin(); power != powers.end(); ++power)
	{
		typedef std::vector<size_t> THist;
		THist hist(nBins, 0);
		for (size_t k = 0; k < numSamples; ++k)
		{
			const TScalar xi = uniform();
			kernel::XYZ chromaticity;
			TScalar pdf;
			const TScalar f = observer.sample(*power, xi, chromaticity, pdf);

			const TBins::const_iterator bin = std::lower_bound(bins.begin(), bins.end(), f);
			if (bin == bins.begin() || bin == bins.end())
			{
				return 1; // f is out of range
			}
			++hist[bin - bins.begin() - 1];
		}

		std::string line;
		std::getline(reference, line);
		std::istringstream buffer(line);
		for (THist::const_iterator count = hist.begin(); count != hist.end(); ++count)
		{
			const TScalar pTest = static_cast<TScalar>(*count) / numSamples;
			TScalar pRef = -1;
			buffer >> pRef;
			if (pRef < 0 || pRef > 1)
			{
				return 1;
			}
			if (num::abs(pRef - pTest) > 1e-3)
			{
				return 1;
			}
		}	
	}

	return 0;
}