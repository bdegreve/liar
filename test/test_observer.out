0.0003657	0.034511600000000003	0.22829099999999999	0.22003700000000001	0.1029978	0.1182144	0.19967579999999999	0.090334399999999995	0.0055465000000000002	2.58e-05
0.00078770000000000001	0.073759099999999994	0.4390442	0.27798270000000003	0.039102400000000002	0.0300284	0.090634699999999999	0.045685200000000002	0.0029589999999999998	1.66e-05
0.000301	0.0297316	0.24610319999999999	0.3809939	0.23802209999999999	0.077084	0.022253599999999998	0.0051969	0.00031270000000000001	9.9999999999999995e-07
4.7999999999999998e-06	0.0	0.0	0.0012719999999999999	0.031998199999999997	0.24743770000000001	0.48554069999999999	0.22034909999999999	0.013333899999999999	6.3600000000000001e-05
0.00030899999999999998	0.0296352	0.21082989999999999	0.24637310000000001	0.13633970000000001	0.1189419	0.17503099999999999	0.077769000000000005	0.0047496999999999999	2.1500000000000001e-05
